File: services/module2b-surrogate/transport/src/ScalarTransport.jl
"""
ScalarTransport — GPU-accelerated steady-state passive scalar transport on structured grids.

Solves the steady advection-diffusion equation for a passive scalar phi:
    U . nabla(phi) = kappa_eff * laplacian(phi)

Designed for FuXi-CFD wind fields (300x300x27 @ 30m).
All operations use CUDA.jl CuArrays for GPU execution.
No scalar indexing — all operations are vectorized broadcast or sliced views.

Reference:
    Ferziger, Peric (2002). Computational Methods for Fluid Dynamics.
    Chapter 3 (Finite Volume) + Chapter 5 (Scalar Transport).
"""
module ScalarTransport

using CUDA

export solve_transport, transport_T_q, build_terrain_mask
export HF_Z_LEVELS, NX, NY, NZ, DX, DY, compute_dz

# ── FuXi-CFD HuggingFace dataset grid specification ──────────────────────────

const HF_Z_LEVELS = Float64[
    5, 10, 15, 20, 25, 30, 35, 40, 45, 50,
    55, 60, 65, 70, 75, 80, 85, 90, 95, 100,
    106.5, 114.95, 125.94, 140.22, 158.78, 182.91, 214.29,
]

const NX = 300
const NY = 300
const NZ = 27
const DX = 30.0  # meters
const DY = 30.0  # meters


"""
    compute_dz(z_levels) -> Vector{Float64}

Cell heights from level centers using midpoint rule.
"""
function compute_dz(z_levels::AbstractVector{Float64})
    n = length(z_levels)
    dz = zeros(Float64, n)
    dz[1] = (z_levels[1] + z_levels[2]) / 2.0
    for k in 2:n-1
        dz[k] = (z_levels[k+1] - z_levels[k-1]) / 2.0
    end
    dz[n] = z_levels[n] - (z_levels[n-1] + z_levels[n]) / 2.0
    return dz
end


# ── Advection (upwind, first order) ─────────────────────────────────────────

"""
    advection_x!(rhs, phi, u, dx)

Compute -d(u*phi)/dx using first-order upwind. In-place on GPU arrays.
Arrays are (nz, ny, nx).
"""
function advection_x!(rhs::CuArray{Float64,3}, phi::CuArray{Float64,3},
                      u::CuArray{Float64,3}, dx::Float64)
    nz, ny, nx = size(phi)

    # Face flux at i+1/2 for i=1..nx-1
    face = CUDA.zeros(Float64, nz, ny, nx - 1)
    @views @. face = ifelse(u[:,:,1:end-1] > 0,
                            u[:,:,1:end-1] * phi[:,:,1:end-1],
                            u[:,:,1:end-1] * phi[:,:,2:end])

    # rhs for interior cells (i=2..nx-1)
    @views @. rhs[:,:,2:end-1] += -(face[:,:,2:end] - face[:,:,1:end-1]) / dx

    return nothing
end

"""
    advection_y!(rhs, phi, v, dy)

Compute -d(v*phi)/dy using first-order upwind.
"""
function advection_y!(rhs::CuArray{Float64,3}, phi::CuArray{Float64,3},
                      v::CuArray{Float64,3}, dy::Float64)
    nz, ny, nx = size(phi)

    face = CUDA.zeros(Float64, nz, ny - 1, nx)
    @views @. face = ifelse(v[:,1:end-1,:] > 0,
                            v[:,1:end-1,:] * phi[:,1:end-1,:],
                            v[:,1:end-1,:] * phi[:,2:end,:])

    @views @. rhs[:,2:end-1,:] += -(face[:,2:end,:] - face[:,1:end-1,:]) / dy
    return nothing
end

"""
    advection_z!(rhs, phi, w, dz)

Compute -d(w*phi)/dz using first-order upwind with non-uniform spacing.
"""
function advection_z!(rhs::CuArray{Float64,3}, phi::CuArray{Float64,3},
                      w::CuArray{Float64,3}, dz_gpu::CuArray{Float64,1})
    nz, ny, nx = size(phi)
    dz3 = reshape(dz_gpu, nz, 1, 1)  # (nz, 1, 1) for broadcasting

    # Face flux at k+1/2 for k=1..nz-1
    face = CUDA.zeros(Float64, nz - 1, ny, nx)
    @views @. face = ifelse(w[1:end-1,:,:] > 0,
                            w[1:end-1,:,:] * phi[1:end-1,:,:],
                            w[1:end-1,:,:] * phi[2:end,:,:])

    # Interior cells (k=2..nz-1)
    @views @. rhs[2:end-1,:,:] += -(face[2:end,:,:] - face[1:end-1,:,:]) / dz3[2:end-1,:,:]

    # Bottom (k=1): only top face
    @views @. rhs[1:1,:,:] += -face[1:1,:,:] / dz3[1:1,:,:]

    # Top (k=nz): only bottom face
    @views @. rhs[end:end,:,:] += face[end:end,:,:] / dz3[end:end,:,:]

    return nothing
end


# ── Diffusion ────────────────────────────────────────────────────────────────

"""
    diffusion!(rhs, phi, dx, dy, dz_gpu, kappa)

Second-order central-difference Laplacian: kappa * nabla^2(phi).
Zero-gradient at z-boundaries.
"""
function diffusion!(rhs::CuArray{Float64,3}, phi::CuArray{Float64,3},
                    dx::Float64, dy::Float64,
                    dz_gpu::CuArray{Float64,1}, kappa::Float64)
    nz, ny, nx = size(phi)
    dz3 = reshape(dz_gpu, nz, 1, 1)

    # x-direction (interior only)
    @views @. rhs[:,:,2:end-1] += kappa * (phi[:,:,3:end] - 2*phi[:,:,2:end-1] + phi[:,:,1:end-2]) / dx^2

    # y-direction
    @views @. rhs[:,2:end-1,:] += kappa * (phi[:,3:end,:] - 2*phi[:,2:end-1,:] + phi[:,1:end-2,:]) / dy^2

    # z-direction (interior)
    @views @. rhs[2:end-1,:,:] += kappa * (phi[3:end,:,:] - 2*phi[2:end-1,:,:] + phi[1:end-2,:,:]) / dz3[2:end-1,:,:]^2

    # z-boundaries: zero-gradient Laplacian (use 1:1 range to avoid scalar indexing)
    @views @. rhs[1:1,:,:] += kappa * (phi[2:2,:,:] - phi[1:1,:,:]) / dz3[1:1,:,:]^2
    @views @. rhs[end:end,:,:] += kappa * (phi[end-1:end-1,:,:] - phi[end:end,:,:]) / dz3[end:end,:,:]^2

    return nothing
end


# ── Boundary conditions ──────────────────────────────────────────────────────

"""
    apply_bcs!(phi, bc3, bc_top, u, v, w, dz0, lapse_rate, terrain_mask)

Phi-based inlet/outlet BCs on all boundaries. All args are CuArrays.
bc3: (nz, 1, 1) profile for broadcasting.
bc_top: (1, 1) scalar broadcast for top BC (avoids scalar indexing on bc_profile[end]).
"""
function apply_bcs!(phi::CuArray{Float64,3},
                    bc3::CuArray{Float64,3},
                    bc_top::CuArray{Float64,2},
                    u::CuArray{Float64,3},
                    v::CuArray{Float64,3},
                    w::CuArray{Float64,3},
                    dz0::Float64,
                    lapse_rate::Float64,
                    terrain_mask::Union{Nothing,CuArray{Bool,3}})
    nz, ny, nx = size(phi)

    # West face (x=1): inflow if u > 0
    @views @. phi[:,:,1] = ifelse(u[:,:,1] > 0, bc3[:,:,1], phi[:,:,2])

    # East face (x=end): inflow if u < 0
    @views @. phi[:,:,end] = ifelse(u[:,:,end] < 0, bc3[:,:,1], phi[:,:,end-1])

    # South face (y=1): inflow if v > 0
    @views @. phi[:,1,:] = ifelse(v[:,1,:] > 0, bc3[:,1,:], phi[:,2,:])

    # North face (y=end): inflow if v < 0
    @views @. phi[:,end,:] = ifelse(v[:,end,:] < 0, bc3[:,1,:], phi[:,end-1,:])

    # Top: Dirichlet for subsidence/calm, Neumann for updraft
    # bc_top is (1,1) CuArray to avoid scalar indexing
    @views @. phi[end:end,:,:] = ifelse(w[end:end,:,:] > 0.01,
                                         phi[end-1:end-1,:,:],
                                         bc_top)

    # Bottom: zero-gradient or lapse rate
    if abs(lapse_rate) < 1e-10
        @views @. phi[1:1,:,:] = phi[2:2,:,:]
    else
        @views @. phi[1:1,:,:] = phi[2:2,:,:] + lapse_rate * dz0
    end

    # Terrain mask: vectorized — for each masked cell, copy from cell above.
    # We process bottom-up so each level sees the updated level above.
    # Use shifted mask+phi pairs (no scalar indexing, each iteration is a broadcast).
    if terrain_mask !== nothing
        for k in 1:nz-1
            # This is a broadcast over (ny, nx) — no scalar indexing
            kp = min(k + 1, nz)
            @views @. phi[k:k,:,:] = ifelse(terrain_mask[k:k,:,:],
                                             phi[kp:kp,:,:],
                                             phi[k:k,:,:])
        end
    end

    return nothing
end


# ── Terrain mask ─────────────────────────────────────────────────────────────

"""
    build_terrain_mask(dem, z_levels) -> CuArray{Bool,3}

Build 3D terrain mask: true where z_level < DEM elevation.
dem: (ny, nx) CPU array in meters ASL. Built on CPU, transferred to GPU.
"""
function build_terrain_mask(dem::AbstractMatrix{<:Real},
                            z_levels::AbstractVector{Float64})
    ny, nx = size(dem)
    nz = length(z_levels)
    # Vectorized on CPU: reshape z to (nz,1,1), broadcast against dem (1,ny,nx)
    z3 = reshape(z_levels, nz, 1, 1)
    dem3 = reshape(Float64.(dem), 1, ny, nx)
    mask = z3 .< dem3  # (nz, ny, nx) BitArray
    return CuArray(mask)
end


# ── Main solver ──────────────────────────────────────────────────────────────

"""
    solve_transport(u, v, w, bc_profile; kwargs...) -> (phi, info)

Solve steady-state passive scalar transport via pseudo-time Jacobi iteration.

All velocity arrays (u, v, w) should be CuArray{Float64,3} of shape (nz, ny, nx).
bc_profile: CuArray{Float64,1} of shape (nz,).

Returns phi (CuArray) and info Dict with convergence data.
"""
function solve_transport(u::CuArray{Float64,3},
                         v::CuArray{Float64,3},
                         w::CuArray{Float64,3},
                         bc_profile::CuArray{Float64,1};
                         dx::Float64=DX,
                         dy::Float64=DY,
                         dz::Vector{Float64}=compute_dz(HF_Z_LEVELS),
                         kappa::Float64=5.0,
                         max_iter::Int=500,
                         tol::Float64=1e-5,
                         lapse_rate::Float64=0.0,
                         terrain_mask::Union{Nothing,CuArray{Bool,3}}=nothing,
                         verbose::Bool=false)
    nz, ny, nx = size(u)
    dz_gpu = CuArray(dz)

    # Precompute BC arrays to avoid scalar indexing in the loop
    bc3 = reshape(bc_profile, nz, 1, 1)  # (nz, 1, 1) for broadcasting
    bc_top = reshape(bc_profile[end:end], 1, 1)  # (1, 1) — no scalar index
    dz0 = HF_Z_LEVELS[2] - HF_Z_LEVELS[1]  # CPU scalar, fine

    # Initialize: broadcast profile everywhere
    phi = repeat(reshape(bc_profile, nz, 1, 1), 1, ny, nx)

    # Stable pseudo-timestep (reduce on GPU, collect to CPU)
    u_max = max(CUDA.@allowscalar(maximum(abs.(u))),
                CUDA.@allowscalar(maximum(abs.(v))),
                CUDA.@allowscalar(maximum(abs.(w))),
                0.01)
    cfl_adv = min(dx, dy, minimum(dz)) / u_max
    cfl_diff = min(dx^2, dy^2, minimum(dz)^2) / (2 * kappa + 1e-30)
    dt = 0.25 * min(cfl_adv, cfl_diff)

    history = Float64[]
    rhs = CUDA.zeros(Float64, nz, ny, nx)

    for it in 1:max_iter
        fill!(rhs, 0.0)

        # Advection + diffusion
        advection_x!(rhs, phi, u, dx)
        advection_y!(rhs, phi, v, dy)
        advection_z!(rhs, phi, w, dz_gpu)
        diffusion!(rhs, phi, dx, dy, dz_gpu, kappa)

        # Update
        phi_new = phi .+ dt .* rhs

        # BCs
        apply_bcs!(phi_new, bc3, bc_top, u, v, w, dz0, lapse_rate, terrain_mask)

        # Convergence check — CUDA.@allowscalar for the single reduction
        change = CUDA.@allowscalar maximum(abs.(phi_new .- phi))
        push!(history, change)
        phi = phi_new

        if verbose && (it % 100 == 0 || change < tol)
            @info "iter $(lpad(it,4)): dmax=$(round(change; sigdigits=3))"
        end

        if change < tol
            verbose && @info "Converged at iteration $it"
            break
        end
    end

    info = Dict(
        "iterations" => length(history),
        "converged" => history[end] < tol,
        "final_residual" => history[end],
        "kappa" => kappa,
        "dt" => dt,
    )

    return phi, info
end


# ── Convenience: T and q transport ───────────────────────────────────────────

"""
    transport_T_q(u, v, w, dem; kwargs...) -> (T, q, info_T, info_q)

Compute T and q fields by passive transport on a FuXi-CFD wind field.
u, v, w: CuArray{Float64,3} (nz, ny, nx).
dem: Matrix{Float64} (ny, nx) — terrain elevation in meters.

Uses ISA standard atmosphere for T and exponential decay for q.
"""
function transport_T_q(u::CuArray{Float64,3},
                       v::CuArray{Float64,3},
                       w::CuArray{Float64,3},
                       dem::AbstractMatrix{<:Real};
                       kappa_T::Float64=5.0,
                       kappa_q::Float64=5.0,
                       max_iter::Int=500,
                       tol::Float64=1e-5,
                       verbose::Bool=false)
    # ISA standard atmosphere temperature profile
    T_profile = CuArray(288.15 .- 0.0065 .* HF_Z_LEVELS)

    # Exponential humidity profile
    q_profile = CuArray(0.008 .* exp.(-HF_Z_LEVELS ./ 2500.0))

    # Terrain mask
    terrain_mask = build_terrain_mask(dem, HF_Z_LEVELS)

    # Solve T
    verbose && @info "Solving T transport..."
    T, info_T = solve_transport(u, v, w, T_profile;
        kappa=kappa_T, lapse_rate=-0.0065,
        terrain_mask=terrain_mask, max_iter=max_iter, tol=tol, verbose=verbose)

    # Solve q
    verbose && @info "Solving q transport..."
    q, info_q = solve_transport(u, v, w, q_profile;
        kappa=kappa_q, lapse_rate=0.0,
        terrain_mask=terrain_mask, max_iter=max_iter, tol=tol, verbose=verbose)
    clamp!(q, 0.0, Inf)

    return T, q, info_T, info_q
end

end # module

File: services/module2b-surrogate/scalar_transport.py
"""
scalar_transport.py — Steady-state passive scalar transport on structured grids.

Solves the steady advection-diffusion equation for a passive scalar φ:

    U·∇φ = κ_eff ∇²φ

on a structured 3D grid (nz × ny × nx) with a fixed velocity field U.
Designed to compute T and q fields from FuXi-CFD wind output (300×300×27 @ 30m).

Discretization options:
  - Advection: upwind (1st order), QUICK (3rd order), TVD (2nd order, flux-limited)
  - Diffusion: central differences (2nd order)
  - Solver: pseudo-time-stepping Jacobi (GPU-friendly)

Boundary conditions:
  - Lateral: Dirichlet from ERA5 interpolated profile
  - Top: Dirichlet (free atmosphere = ERA5)
  - Bottom (terrain): zero-flux (∂φ/∂n = 0) or specified gradient (lapse rate)

Reference:
    Ferziger, Perić (2002). Computational Methods for Fluid Dynamics.
    Springer. Chapter 3 (Finite Volume) + Chapter 5 (Scalar Transport).
"""
from __future__ import annotations

from enum import Enum
from typing import NamedTuple

import numpy as np

try:
    import torch
    HAS_TORCH = True
except ImportError:
    HAS_TORCH = False


class AdvectionScheme(Enum):
    UPWIND = "upwind"       # 1st order, very diffusive but unconditionally stable
    QUICK = "quick"         # 3rd order (Leonard 1979), good balance accuracy/stability
    TVD_VANLEER = "vanleer" # 2nd order TVD, van Leer limiter — no spurious oscillations


# ── FuXi-CFD grid specification ─────────────────────────────────────────────

# 27 vertical levels (non-uniform spacing, AGL in meters)
FUXI_Z_LEVELS = np.array([
    5, 10, 15, 20, 25, 30, 35, 40, 45, 50,
    55, 60, 65, 70, 80, 90, 100, 110, 120, 130,
    140, 150, 160, 170, 180, 195, 214,
], dtype=np.float64)

FUXI_NX, FUXI_NY, FUXI_NZ = 300, 300, 27
FUXI_DX, FUXI_DY = 30.0, 30.0  # meters


def compute_dz(z_levels: np.ndarray) -> np.ndarray:
    """Compute cell heights from level centers.

    Uses midpoint rule: dz[k] = distance between midpoints of cells k-1/k and k/k+1.
    """
    n = len(z_levels)
    dz = np.zeros(n)

    # First cell: from ground (z=0) to midpoint between level 0 and 1
    dz[0] = (z_levels[0] + z_levels[1]) / 2.0

    # Interior cells
    for k in range(1, n - 1):
        dz[k] = (z_levels[k + 1] - z_levels[k - 1]) / 2.0

    # Last cell: symmetric about last level
    dz[-1] = z_levels[-1] - (z_levels[-2] + z_levels[-1]) / 2.0

    return dz


# ── Advection flux functions ────────────────────────────────────────────────

def _flux_upwind(phi_L: np.ndarray, phi_R: np.ndarray,
                 vel: np.ndarray) -> np.ndarray:
    """First-order upwind flux at face between L and R cells."""
    return np.where(vel > 0, vel * phi_L, vel * phi_R)


def _van_leer_limiter(r: np.ndarray) -> np.ndarray:
    """Van Leer flux limiter: ψ(r) = (r + |r|) / (1 + |r|)."""
    return (r + np.abs(r)) / (1.0 + np.abs(r) + 1e-30)


def _advection_x(phi: np.ndarray, u: np.ndarray, dx: float,
                 scheme: AdvectionScheme) -> np.ndarray:
    """Compute -∂(u·φ)/∂x using the specified scheme."""
    if scheme == AdvectionScheme.UPWIND:
        flux = _flux_upwind(phi, np.roll(phi, -1, axis=2), u)
        flux_m = _flux_upwind(np.roll(phi, 1, axis=2), phi, np.roll(u, 1, axis=2))
        return -(flux - flux_m) / dx

    elif scheme == AdvectionScheme.TVD_VANLEER:
        # Face values with TVD limiter
        # Forward differences
        dphi_fwd = np.roll(phi, -1, axis=2) - phi
        dphi_bwd = phi - np.roll(phi, 1, axis=2)
        dphi_2fwd = np.roll(phi, -2, axis=2) - np.roll(phi, -1, axis=2)

        # Ratio for limiter (upwind-ratio)
        r_pos = dphi_bwd / (dphi_fwd + 1e-30)  # when u > 0
        r_neg = dphi_2fwd / (dphi_fwd + 1e-30)  # when u < 0 (reverse direction)

        psi_pos = _van_leer_limiter(r_pos)
        psi_neg = _van_leer_limiter(r_neg)

        # Face value (TVD reconstruction)
        phi_face_pos = phi + 0.5 * psi_pos * dphi_fwd           # u > 0
        phi_face_neg = np.roll(phi, -1, axis=2) - 0.5 * psi_neg * dphi_fwd  # u < 0

        phi_face = np.where(u > 0, phi_face_pos, phi_face_neg)
        flux = u * phi_face

        # Same for left face
        dphi_fwd_m = phi - np.roll(phi, 1, axis=2)
        dphi_bwd_m = np.roll(phi, 1, axis=2) - np.roll(phi, 2, axis=2)
        dphi_2fwd_m = np.roll(phi, -1, axis=2) - phi

        r_pos_m = dphi_bwd_m / (dphi_fwd_m + 1e-30)
        r_neg_m = dphi_2fwd_m / (dphi_fwd_m + 1e-30)

        u_m = np.roll(u, 1, axis=2)
        phi_face_pos_m = np.roll(phi, 1, axis=2) + 0.5 * _van_leer_limiter(r_pos_m) * dphi_fwd_m
        phi_face_neg_m = phi - 0.5 * _van_leer_limiter(r_neg_m) * dphi_fwd_m
        phi_face_m = np.where(u_m > 0, phi_face_pos_m, phi_face_neg_m)
        flux_m = u_m * phi_face_m

        return -(flux - flux_m) / dx

    elif scheme == AdvectionScheme.QUICK:
        # QUICK: 3-point upstream-weighted (Leonard 1979)
        # For u > 0: φ_face = 3/8 φ_R + 6/8 φ_C - 1/8 φ_L
        phi_C = phi
        phi_R = np.roll(phi, -1, axis=2)
        phi_L = np.roll(phi, 1, axis=2)
        phi_LL = np.roll(phi, 2, axis=2)
        phi_RR = np.roll(phi, -2, axis=2)

        face_pos = (3 * phi_R + 6 * phi_C - phi_L) / 8.0
        face_neg = (3 * phi_C + 6 * phi_R - phi_RR) / 8.0
        phi_face = np.where(u > 0, face_pos, face_neg)
        flux = u * phi_face

        u_m = np.roll(u, 1, axis=2)
        face_pos_m = (3 * phi_C + 6 * phi_L - phi_LL) / 8.0
        face_neg_m = (3 * phi_L + 6 * phi_C - phi_R) / 8.0
        phi_face_m = np.where(u_m > 0, face_pos_m, face_neg_m)
        flux_m = u_m * phi_face_m

        return -(flux - flux_m) / dx


def _advection_y(phi: np.ndarray, v: np.ndarray, dy: float,
                 scheme: AdvectionScheme) -> np.ndarray:
    """Compute -∂(v·φ)/∂y — transpose of x logic."""
    # Swap axes 1<->2 to reuse _advection_x logic
    phi_t = np.swapaxes(phi, 1, 2)
    v_t = np.swapaxes(v, 1, 2)
    result = _advection_x(phi_t, v_t, dy, scheme)
    return np.swapaxes(result, 1, 2)


def _advection_z(phi: np.ndarray, w: np.ndarray, dz: np.ndarray,
                 scheme: AdvectionScheme) -> np.ndarray:
    """Compute -∂(w·φ)/∂z with non-uniform spacing.

    Uses upwind with explicit z-indexing (no np.roll wrap-around).
    """
    nz = phi.shape[0]
    result = np.zeros_like(phi)

    for k in range(1, nz - 1):
        # Upwind flux at top face of cell k
        flux_top = np.where(w[k] > 0, w[k] * phi[k], w[k] * phi[k + 1])
        # Upwind flux at bottom face of cell k
        flux_bot = np.where(w[k - 1] > 0, w[k - 1] * phi[k - 1], w[k - 1] * phi[k])
        result[k] = -(flux_top - flux_bot) / dz[k]

    # Bottom (k=0): only top face
    flux_top_0 = np.where(w[0] > 0, w[0] * phi[0], w[0] * phi[1])
    result[0] = -flux_top_0 / dz[0]

    # Top (k=nz-1): only bottom face
    flux_bot_top = np.where(w[-2] > 0, w[-2] * phi[-2], w[-2] * phi[-1])
    result[-1] = flux_bot_top / dz[-1]

    return result


# ── Diffusion ────────────────────────────────────────────────────────────────

def _diffusion(phi: np.ndarray, dx: float, dy: float, dz: np.ndarray,
               kappa: float) -> np.ndarray:
    """Second-order central-difference Laplacian: κ∇²φ.

    Uses zero-gradient (Neumann) at z-boundaries to avoid wrap-around artifacts.
    """
    nz = phi.shape[0]
    dz3 = dz[:, None, None]

    lap_x = (np.roll(phi, -1, axis=2) - 2 * phi + np.roll(phi, 1, axis=2)) / dx**2
    lap_y = (np.roll(phi, -1, axis=1) - 2 * phi + np.roll(phi, 1, axis=1)) / dy**2

    # z-direction: explicit indexing to avoid wrap
    lap_z = np.zeros_like(phi)
    for k in range(1, nz - 1):
        lap_z[k] = (phi[k + 1] - 2 * phi[k] + phi[k - 1]) / dz[k]**2
    # Boundaries: zero-gradient (∂²φ/∂z² = 0)
    lap_z[0] = (phi[1] - phi[0]) / dz[0]**2
    lap_z[-1] = (phi[-2] - phi[-1]) / dz[-1]**2

    return kappa * (lap_x + lap_y + lap_z)


# ── Boundary conditions ─────────────────────────────────────────────────────

def _apply_bcs(phi: np.ndarray, bc_profile: np.ndarray,
               terrain_mask: np.ndarray | None = None,
               lapse_rate: float = 0.0) -> None:
    """Apply boundary conditions in-place.

    Lateral + top: Dirichlet from ERA5 profile.
    Bottom: zero-gradient (if lapse_rate=0) or imposed gradient.
    Terrain: zero-flux (copy from cell above).
    """
    nz = phi.shape[0]

    # Lateral Dirichlet (all 4 faces)
    phi[:, :, 0] = bc_profile[:, None]
    phi[:, :, -1] = bc_profile[:, None]
    phi[:, 0, :] = bc_profile[:, None]
    phi[:, -1, :] = bc_profile[:, None]

    # Top: Dirichlet
    phi[-1, :, :] = bc_profile[-1]

    # Bottom: zero-gradient or lapse rate
    if abs(lapse_rate) < 1e-10:
        phi[0, :, :] = phi[1, :, :]
    else:
        dz0 = FUXI_Z_LEVELS[1] - FUXI_Z_LEVELS[0]
        phi[0, :, :] = phi[1, :, :] + lapse_rate * dz0

    # Terrain mask: copy from cell directly above
    if terrain_mask is not None:
        for k in range(1, nz):
            phi[k][terrain_mask[k]] = phi[min(k + 1, nz - 1)][terrain_mask[k]]
        phi[0][terrain_mask[0]] = phi[1][terrain_mask[0]]


# ── Main solver ──────────────────────────────────────────────────────────────

def solve_scalar_transport(
    u: np.ndarray,
    v: np.ndarray,
    w: np.ndarray,
    bc_profile: np.ndarray,
    dx: float = FUXI_DX,
    dy: float = FUXI_DY,
    dz: np.ndarray | None = None,
    kappa: float = 5.0,
    terrain_mask: np.ndarray | None = None,
    lapse_rate: float = 0.0,
    scheme: str = "vanleer",
    max_iter: int = 500,
    tol: float = 1e-5,
    verbose: bool = False,
) -> tuple[np.ndarray, dict]:
    """Solve steady-state passive scalar transport.

    Args:
        u, v, w: velocity field (nz, ny, nx) from FuXi-CFD or CFD [m/s]
        bc_profile: boundary profile (nz,) — T(z) in K or q(z) in kg/kg
        dx, dy: horizontal grid spacing [m] (default 30m)
        dz: vertical spacing (nz,), computed from FUXI_Z_LEVELS if None
        kappa: turbulent diffusivity [m²/s] (~5-10 for ABL thermal)
        terrain_mask: (nz, ny, nx) or (ny, nx) bool — True inside terrain.
            If 2D, expanded to 3D assuming terrain at lowest levels.
        lapse_rate: dT/dz at ground [K/m], 0 = adiabatic (default)
        scheme: "upwind", "vanleer", "quick"
        max_iter: maximum pseudo-time iterations
        tol: convergence tolerance on max(|δφ|)
        verbose: print iteration info

    Returns:
        (phi, info): scalar field (nz, ny, nx) and convergence info dict
    """
    nz, ny, nx = u.shape
    adv_scheme = AdvectionScheme(scheme)

    if dz is None:
        dz = compute_dz(FUXI_Z_LEVELS[:nz])

    # Handle 2D terrain mask → 3D
    if terrain_mask is not None and terrain_mask.ndim == 2:
        # Assume terrain mask means "ground is present" at (y, x)
        # No 3D mask available without DEM — skip interior masking
        terrain_3d = None
    elif terrain_mask is not None and terrain_mask.ndim == 3:
        terrain_3d = terrain_mask
    else:
        terrain_3d = None

    # Initialize: inflow profile everywhere
    phi = np.broadcast_to(bc_profile[:, None, None], (nz, ny, nx)).copy().astype(np.float64)

    # Estimate stable pseudo-timestep
    u_max = max(np.abs(u).max(), np.abs(v).max(), np.abs(w).max(), 0.01)
    cfl_adv = min(dx, dy, float(dz.min())) / u_max
    cfl_diff = min(dx**2, dy**2, float(dz.min())**2) / (2 * kappa + 1e-30)
    dt = 0.25 * min(cfl_adv, cfl_diff)

    history = []

    for it in range(max_iter):
        # Advection
        rhs = _advection_x(phi, u, dx, adv_scheme)
        rhs += _advection_y(phi, v, dy, adv_scheme)
        rhs += _advection_z(phi, w, dz, adv_scheme)

        # Diffusion
        rhs += _diffusion(phi, dx, dy, dz, kappa)

        # Update
        phi_new = phi + dt * rhs

        # BCs
        _apply_bcs(phi_new, bc_profile, terrain_3d, lapse_rate)

        # Convergence
        change = np.abs(phi_new - phi).max()
        history.append(change)
        phi = phi_new

        if verbose and (it % 100 == 0 or change < tol):
            print(f"  iter {it:4d}: Δmax={change:.2e}, "
                  f"range=[{phi.min():.4f}, {phi.max():.4f}]")

        if change < tol:
            if verbose:
                print(f"  Converged at iteration {it}")
            break

    info = {
        "iterations": it + 1,
        "converged": change < tol,
        "final_residual": float(change),
        "history": history,
        "scheme": scheme,
        "kappa": kappa,
        "dt": dt,
    }

    return phi, info


# ── Convenience: process a FuXi-CFD case ─────────────────────────────────────

def transport_T_q_on_wind_field(
    u: np.ndarray,           # (27, 300, 300)
    v: np.ndarray,
    w: np.ndarray,
    T_profile: np.ndarray,   # (27,) ERA5 temperature [K]
    q_profile: np.ndarray,   # (27,) ERA5 specific humidity [kg/kg]
    terrain_mask: np.ndarray | None = None,
    kappa_T: float = 5.0,
    kappa_q: float = 5.0,
    lapse_rate_T: float = -0.0065,  # dry adiabatic lapse rate [K/m]
    scheme: str = "vanleer",
    max_iter: int = 500,
    tol: float = 1e-5,
    verbose: bool = False,
) -> dict:
    """Compute T and q fields by passive transport on a wind field.

    Args:
        u, v, w: wind components (nz, ny, nx) [m/s]
        T_profile: ERA5 temperature profile (nz,) [K]
        q_profile: ERA5 specific humidity profile (nz,) [kg/kg]
        terrain_mask: (ny, nx) or (nz, ny, nx) bool
        kappa_T: thermal diffusivity [m²/s] (5-10 typical for ABL)
        kappa_q: moisture diffusivity [m²/s] (≈ kappa_T)
        lapse_rate_T: surface temperature gradient [K/m] (-0.0065 = dry adiabatic)
        scheme: advection scheme ("upwind", "vanleer", "quick")

    Returns:
        dict with T (nz, ny, nx) [K], q (nz, ny, nx) [kg/kg], info_T, info_q.
    """
    if verbose:
        print("Solving T transport...")
    T, info_T = solve_scalar_transport(
        u, v, w, T_profile,
        kappa=kappa_T, terrain_mask=terrain_mask,
        lapse_rate=lapse_rate_T, scheme=scheme,
        max_iter=max_iter, tol=tol, verbose=verbose,
    )

    if verbose:
        print(f"  T: [{T.min():.1f}, {T.max():.1f}] K, {info_T['iterations']} iter")
        print("Solving q transport...")

    q, info_q = solve_scalar_transport(
        u, v, w, q_profile,
        kappa=kappa_q, terrain_mask=terrain_mask,
        lapse_rate=0.0, scheme=scheme,
        max_iter=max_iter, tol=tol, verbose=verbose,
    )
    q = np.clip(q, 0, None)

    if verbose:
        print(f"  q: [{q.min():.6f}, {q.max():.6f}] kg/kg, {info_q['iterations']} iter")

    return {"T": T, "q": q, "info_T": info_T, "info_q": info_q}
